python/tests/test_ld_matrix.py

```diff
@@ -0,0 +1,559 @@
+
+    while r < len(row_sites) and c < len(col_sites):
+        if row_sites[r] < col_sites[c]:
+            idx.rdiff_sites.append(s)
+            idx.rdiff_matrix.append(r)
+            idx.sites.append(row_sites[r])
+            r += 1
+            s += 1
+        elif col_sites[c] < row_sites[r]:
+            idx.cdiff_sites.append(s)
+            idx.cdiff_matrix.append(c)
+            idx.sites.append(col_sites[c])
+            c += 1
+            s += 1
+        else:
+            idx.rshr_matrix.append(r)
+            idx.cshr_matrix.append(c)
+            idx.shr_sites.append(s)
+            idx.sites.append(row_sites[r])
+            r += 1
+            c += 1
+            s += 1
+
+    # Once we've hit the end of one (or both) of the row/column arrays,
+    # check to see if there are any remaining items in any of the lists.
+    # If there are, account for the remaining values.
+    while r < len(row_sites):
+        idx.rdiff_matrix.append(r)
+        idx.rdiff_sites.append(s)
+        idx.sites.append(row_sites[r])
+        r += 1
+    while c < len(col_sites):
+        idx.cdiff_matrix.append(c)
+        idx.cdiff_sites.append(s)
+        idx.sites.append(col_sites[c])
+        c += 1
+
+    return idx
+
+
+def compute_two_site_general_stat(
+    state,
+    func,
+    polarized,
+    norm_method,
+    sample_sets,
+    idx,
+    debug=False,
+    print_weights=False,
+):
+    """Compute the ld matrix, given the sample allele state, various parameters,
+    and a site index.
+
+    :param state: State matrix containing the allelic state for each sample
+    :param func: Summary function to apply to the haplotype weights
+    :param polarized: If true, skip the computation of the statistic for the
+                      ancestral state.
+    :param norm_method: Method of producing the normalizing constants (see the
+                        docstring for the NormMethod enum).
+    :param sample_sets: List of lists of samples to compute stats for. We will
+                        only produce haplotype matricies that contain these
+                        samples, resulting in stats that are computed on subsets
+                        of the total number of samples on the tree sequence.
+    :param idx: Site index, storing indices for the state matrix, result matrix,
+                and site ids.
+    :param debug: If true, print the haplotype matrix and normalizing constants
+    :param print_weights: Debugging tool for printing out the computed haplotype
+                          frequencies.
+    :returns: 3d numpy array containing LD for (sample_set,row_site,column_site)
+    """
+    state = np.asarray(state)
+    norm = NormMethod(norm_method)
+
+    result = np.zeros((len(sample_sets), idx.n_rows, idx.n_cols))
+
+    def compute(left_states, right_states):
+        hap_counts = np.zeros((np.max(left_states) + 1, np.max(right_states) + 1))
+        # compute the haplotype matrix
```

Comment:
This is still confusing me - what's the haplotype matrix? (I'm pretty sure it's not the same as what we get when calling ``ts.haplotypes()``, e.g.). Can we rename this please?